File: app/src/main/java/com/psliusar/layers/sample/screen/dialog/DialogsLayer.java
package com.psliusar.layers.sample.screen.dialog;

import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.psliusar.layers.DialogWrapper;
import com.psliusar.layers.Layer;
import com.psliusar.layers.binder.Bind;
import com.psliusar.layers.sample.R;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class DialogsLayer extends Layer<DialogsViewModel> implements View.OnClickListener,
        CustomDialogLayer.OnCustomDialogListener {

    private static final String DIALOG_SIMPLE = "SimpleDialog";
    private static final String DIALOG_CUSTOM = "CustomDialog";

    @Bind(value = R.id.dialogs_simple, clicks = true) View simpleButton;
    @Bind(value = R.id.dialogs_custom, clicks = true) View customButton;

    @Nullable
    @Override
    protected DialogsViewModel onCreateViewModel() {
        return new DialogsViewModel();
    }

    @Nullable
    @Override
    protected View onCreateView(@Nullable ViewGroup parent) {
        return inflate(R.layout.screen_dialogs, parent);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
        case R.id.dialogs_simple:
            getViewModel().simpleDialogClick(this);
            break;
        case R.id.dialogs_custom:
            getViewModel().customDialogClick(this);
            break;
        }
    }

    @Override
    public void onDialogAction1(@NonNull CustomDialogLayer layer) {
        if (DIALOG_CUSTOM.equals(layer.getName())) {
            Toast.makeText(getContext(), "Action 1 was selected", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onDialogAction2(@NonNull CustomDialogLayer layer) {
        if (DIALOG_CUSTOM.equals(layer.getName())) {
            Toast.makeText(getContext(), "Action 2 was selected", Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onDialogCancel(@NonNull DialogWrapper wrapper) {
        Toast.makeText(getContext(), "Dialog " + DIALOG_SIMPLE + " was dismissed", Toast.LENGTH_SHORT).show();
    }

    void showSimpleDialog(String title, String message) {
        final Bundle args = SimpleDialogLayer.createArguments(title, message);
        getLayers().add(SimpleDialogLayer.class)
                .setArguments(args)
                .setName(DIALOG_SIMPLE)
                .commit();
    }

    void showCustomDialog(String title) {
        final Bundle args = CustomDialogLayer.createArguments(title);
        getLayers().add(CustomDialogLayer.class)
                .setArguments(args)
                .setName(DIALOG_CUSTOM)
                .commit();
    }
}

File: app/src/main/java/com/psliusar/layers/sample/screen/dialog/DialogsLayer.kt
package com.psliusar.layers.sample.screen.dialog

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.psliusar.layers.DialogWrapper
import com.psliusar.layers.Layer
import com.psliusar.layers.sample.R

private const val DIALOG_SIMPLE = "SimpleDialog"
private const val DIALOG_CUSTOM = "CustomDialog"

class DialogsLayer : Layer(), CustomDialogLayer.OnCustomDialogListener {

    override fun onCreateView(savedState: Bundle?, parent: ViewGroup?): View? {
        return inflate(R.layout.screen_dialogs, parent)
    }

    override fun onBindView(savedState: Bundle?, view: View) {
        super.onBindView(savedState, view)
        view.findViewById<View>(R.id.dialogs_simple).setOnClickListener {
            showSimpleDialog("Hello World!", "This is simple AlertDialog controlled via Layer")
        }
        view.findViewById<View>(R.id.dialogs_custom).setOnClickListener {
            showCustomDialog("Custom dialog")
        }
    }

    override fun onDialogAction1(dialog: CustomDialogLayer) {
        if (dialog.name == DIALOG_CUSTOM) {
            Toast.makeText(context, "Action 1 was selected", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onDialogAction2(dialog: CustomDialogLayer) {
        if (dialog.name == DIALOG_CUSTOM) {
            Toast.makeText(context, "Action 2 was selected", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onDialogCancel(wrapper: DialogWrapper) {
        Toast.makeText(context, "Dialog $DIALOG_SIMPLE was dismissed", Toast.LENGTH_SHORT).show()
    }

    private fun showSimpleDialog(title: String, message: String) {
        layers.add<SimpleDialogLayer> {
            arguments = SimpleDialogLayer.createArguments(title, message)
            name = DIALOG_SIMPLE
        }
    }

    private fun showCustomDialog(title: String) {
        layers.add<CustomDialogLayer> {
            arguments = CustomDialogLayer.createArguments(title)
            name = DIALOG_CUSTOM
        }
    }
}
